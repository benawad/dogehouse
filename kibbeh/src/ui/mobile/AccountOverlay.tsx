import React, { useEffect, useRef, ReactNode } from "react";
import Draggable, {
  DraggableBounds,
  ControlPosition,
  DraggableEventHandler,
} from "react-draggable";

export interface AccountOverlyProps {
  children: ReactNode;
  position: ControlPosition;
  dragHandler: DraggableEventHandler;
  closeHandler: () => void;
  bounds: DraggableBounds;
  className?: string;
  axis?: "both" | "x" | "y" | "none" | undefined;
}

export const AccountOverlay: React.FC<AccountOverlyProps> = ({
  children,
  position,
  bounds,
  dragHandler,
  closeHandler,
  className = "",
  axis = "y",
}) => {
  const endPosition = useRef(0);
  const node = useRef<HTMLDivElement>(null);

  useEffect(() => {
    const handleClick = (e: MouseEvent) => {
      if (!(node!.current! as Node).contains(e.target as Node)) {
        endPosition.current = 0;
        closeHandler();
      }
    };

    document.addEventListener("mousedown", handleClick);

    return () => {
      document.removeEventListener("mousedown", handleClick);
    };
  }, [closeHandler]);

  const onStop = () => {
    const { y } = position;
    const { current } = endPosition;

    if (y > current + 15) closeHandler();

    endPosition.current = y;
  };

  return (
    <Draggable
      onDrag={dragHandler}
      onStop={onStop}
      axis={axis}
      handle="#handle"
      bounds={bounds}
      position={position}
    >
      <div
        ref={node}
        style={{ padding: "43px 25px" }}
        className={`w-full rounded-t-8 px-2 pt-7 pb-5 bg-primary-800 relative  ${className}`}
        data-testid="account-overlay"
      >
        <span
          id="handle"
          style={{ cursor: "grab" }}
          className="absolute bg-primary-300 rounded-8 inline-block top-3 left-1 right-1 w-6 h-1 m-auto"
        ></span>
        {children}
      </div>
    </Draggable>
  );
};
