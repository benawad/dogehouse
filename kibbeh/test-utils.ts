import { render } from "@testing-library/react";
import "@testing-library/jest-dom";
import "mutationobserver-shim";

const customRender = (ui: any, options = {}) =>
  render(ui, {
    ...options,
  });

export * from "@testing-library/react";

export { customRender as render };
