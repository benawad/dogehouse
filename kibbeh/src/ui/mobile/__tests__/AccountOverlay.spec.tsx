// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-nocheck @todo this file needs to be fixed

import React from "react";

import { render } from "../../../../test-utils";
import { AccountOverlay } from "../AccountOverlay";

describe("AccountOverlay", () => {
  it("should render correctly", () => {
    const { getByTestId, getByText } = render(
      <AccountOverlay>
        <div className="flex">Child</div>
      </AccountOverlay>
    );
    const component = getByTestId("account-overlay");
    const child = getByText("Child");

    expect(component).toBeVisible();
    expect(child).toBeVisible();
  });

  it("should match snapshot", () => {
    const { getByTestId } = render(
      <AccountOverlay>
        <div className="flex">Child</div>
      </AccountOverlay>
    );
    const component = getByTestId("account-overlay");

    expect(component).toMatchSnapshot();
  });
});
