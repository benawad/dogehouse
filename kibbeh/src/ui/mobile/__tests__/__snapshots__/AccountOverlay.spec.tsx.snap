// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountOverlay should match snapshot 1`] = `
<div
  class="w-full rounded-t-8 px-2 pt-7 pb-5 bg-primary-800 relative   react-draggable"
  data-testid="account-overlay"
  style="padding: 43px 25px; transform: translate(0px,0px);"
>
  <span
    class="absolute bg-primary-300 rounded-8 inline-block top-3 left-1 right-1 w-6 h-1 m-auto"
    id="handle"
    style="cursor: grab;"
  />
  <div
    class="flex"
  >
    Child
  </div>
</div>
`;
